File: leetcodes_devrod/arrays/array_except_self/array_except_self.ts

// Solução "força bruta" O(n^2) - não é a mais eficiente, mas é a mais simples de entender

export const arrayExeptSelf = (nums: number[]): number[] => {
    const result: number[]  = new Array(nums.length).fill(1);
    let left = 1;
    for(let i  = 0; i < nums.length; i++){
        result[i] = left;
        left *= nums[i];
    }   
    let right = 1;
    for(let i = nums.length - 1; i >= 0; i--){
        result[i] *= right;
        right *= nums[i];
    }   
    return result;
}

// Teste da função arrayExeptSelf
function testArrayExceptSelf() {
    const testCases = [
        {input: [1, 2, 3, 4], expected: [24, 12, 8, 6]},
        {input: [5, 6, 7], expected: [42, 35, 30]},
        {input: [1], expected: [1]},
        {input: [0], expected: [0]},
    ]

    for (const testCase of testCases) {
        const result = arrayExeptSelf(testCase.input);
        if (JSON.stringify(result) === JSON.stringify(testCase.expected)) {
            console.log(`Test case ${JSON.stringify(testCase.input)} passed.`);
        } else {
            console.error(`Test case ${JSON.stringify(testCase.input)} failed. Expected ${testCase.expected}, but got ${result}.`);
        }
    }
}

// Versão otimizada O(n) - mais eficiente, mas um pouco mais complexa de entender


export const arrayExceptSelfOptimized = (nums: number[]): number[] => {

    const res: number[] = new Array(nums.length).fill(1);
    let left = 1;

    for (let i = 0; i < nums.length; i++) {
        res[i] = left;
        left *= nums[i];
    }
    let right = 1;

    for (let i = nums.length - 1; i >= 0; i--) {
        res[i] *= right;
        right *= nums[i];
    }
    return res;
}


// Teste da função arrayExceptSelfOptimized
function testArrayExceptSelfOptimized() {
    const testCases = [
        {input: [1, 2, 3, 4], expected: [24, 12, 8, 6]},
        {input: [5, 6, 7], expected: [42, 35, 30]},
        {input: [1], expected: [1]},
        {input: [0], expected: [0]},
    ]

    for (const testCase of testCases) {
        const result = arrayExceptSelfOptimized(testCase.input);
        if (JSON.stringify(result) === JSON.stringify(testCase.expected)) {
            console.log(`Test case ${JSON.stringify(testCase.input)} passed.`);
        } else {
            console.error(`Test case ${JSON.stringify(testCase.input)} failed. Expected ${testCase.expected}, but got ${result}.`);
        }
    }
}


// Versão otimizada O(n) - mais eficiente, mas um pouco mais complexa de entender
export const arrayExceptSelfOptimized2 = (nums: number[]): number[] => {
   const res:number[] = new Array(nums.length).fill(1);
   nums.reduce(
    (acc, curr, i) => (res[i] *= acc, acc * curr),
    1
    );

    nums.reduceRight(
        (acc, curr, i) => (res[i] *= acc, acc * curr),
        1
    );
    return res;
}

// Teste da função arrayExceptSelfOptimized2
export const testArrayExceptSelfOptimized2 = () => {
    const target = [
        {input: [1, 2, 3, 4], expected: [24, 12, 8, 6]},
        {input: [5, 6, 7], expected: [42, 35, 30]},
        {input: [1], expected: [1]},
        {input: [0], expected: [0]},
    ]

    target.forEach((testCase) => {
        const result  = arrayExceptSelfOptimized2(testCase.input);
        if (JSON.stringify(result) === JSON.stringify(testCase.expected)) {
            console.log(`Test case ${JSON.stringify(testCase.input)} passed.`);
        }
        else {
            console.error(`Test case ${JSON.stringify(testCase.input)} failed. Expected ${testCase.expected}, but got ${result}.`);
        }
    })
    console.log("All test cases passed.");  
    return true;
}

File: leetcodes_devrod/arrays/array_except_self/ArrayExceptSelf.java
package leetcodes_devrod.arrays.array_except_self;

public class ArrayExceptSelf {


    public static int[] productExceptSelf(int[] nums){
        int[] res = new int[nums.length];
        int n = nums.length;
        int left = 1;

        int right = 1;
        for(int i = 0; i < n; i++){
            res[i] = left;
            left *= nums[i];
        }
        for(int i = n - 1; i >= 0; i--){
            res[i] *= right;
            right *= nums[i];
        }
        return res;
    }

    public static void main(String[] args) throws Exception {
        int[] nums = {1, 2, 3, 4};
        int[] result = productExceptSelf(nums);
        System.out.print("Input: " + java.util.Arrays.toString(nums) + "\nOutput: " + java.util.Arrays.toString(result));
        // Output será esse aqui ninjas: [24, 12, 8, 6]   
    }


    //Bora testar?
    public static void testArrayExceptSelf(){
        int[] nums = {1, 2, 3, 4};
        int[] expected = {24, 12, 8, 6};
        int[] result = productExceptSelf(nums);
        
        if(java.util.Arrays.equals(result, expected)){
            System.out.println("Test passed!");
        } else {
            System.out.println("Test failed. Expected: " + java.util.Arrays.toString(expected) + ", but got: " + java.util.Arrays.toString(result));
        }
    }
}
